### ISSUE TITLE
Scores Histogram

### ISSUE THREAD

**Issue by hamid13r** [SBC-Utrecht/pytom-template-matching-gpu#45]
I have been trying pytom template matching on various tomograms and cannot use the automatic thresholding of the scores. It seems like the problem is no matter how many candidates I extract I don't get a high number of "false" positives so the code can fit the curve properly. Even with a low threshold and more than 10K candidates I just get a Gaussian histogram. 

Any idea what am I doing wrong? This happens with and without applying amira masks on my tomogram. 

For cellular tomography:
![image](https://github.com/SBC-Utrecht/pytom-template-matching-gpu/assets/66981513/b3f8c6bc-acaf-4fc4-9dc2-1517a7d95100)

For tomogram from isolated proteins:
![image](https://github.com/SBC-Utrecht/pytom-template-matching-gpu/assets/66981513/e251dfa5-537a-4ce6-9673-9c06edd9925b)

**Comment by McHaillet** [comment]
Hi Hamid,

I have been on holidays the last weeks, so only now am looking at your question.

Whether you can separate the particles from the background will heavily depend on the contrast. So to answer that I will need some more info:
- what is the number of particles you expect in the data? for the histogram fit I usually as a rule-of-thumb extract 4 times the expected number of particles?
- what is the diameter of your particles along the largest dimensions? (and related what is the pixel size of tomograms, and the angular sampling?)

Alternatively I am also happy to have a call with you to discuss it!

Best,
Marten

**Comment by hamid13r** [comment]
Marten, 

Thank you for your response. My tomograms are 9.98A/pix and the sample is a 70S ribosome so I used a radius of 15 pixels. Seems like this is happening no matter how high of a number I choose since I still get this histogram when I extract every possible candidate (like I put 20K and it runs out after ~6K). 

I wanted to try it with the test tomograms from your paper but I only saw the tilt-series and I think it will take me a while to install and run aretomo.

Also, the sampling is 7 degrees. 

Best,
Hamid

**Comment by McHaillet** [comment]
Hi Hamid!

> Thank you for your response. My tomograms are 9.98A/pix and the sample is a 70S ribosome so I used a radius of 15 pixels. Seems like this is happening no matter how high of a number I choose since I still get this histogram when I extract every possible candidate (like I put 20K and it runs out after ~6K).

I think after 6K the score map is just empty because I mask out spheres, based on the extraction radius, around each location. For the histogram fit I would rather just recommend extracting 3-4 times the number of expected particles in your tomogram (so for you 500?). Otherwise the background becomes to dominant in the fitting. In your case it looks like this is the ribosome population (blue line):

<img src="https://github.com/SBC-Utrecht/pytom-template-matching-gpu/assets/58044494/1eb0f968-f85e-4ff3-8bbb-8d972b3cb198" width="400" height="300">

For the 10 A/pix tomograms you might also need higher angular sampling, 180 / (pi * (1 / 20A) * 300A) = 3.8° . (but needless to say that might take quite long, so depends on whether that is feasible for your project)


> I wanted to try it with the test tomograms from your paper but I only saw the tilt-series and I think it will take me a while to install and run aretomo.

I uploaded a dataset of 10 test tomograms from the paper for which I will write a tutorial (on the to-do list). So if you want to benchmark on those, feel free to already download the dataset. The methods in the paper should be comprehensive for running it.

Cheers!

**Comment by hamid13r** [comment]
Thank you very much. The problem is the number of particles comes out ~10-20 when I let the program fit the distribution (no matter what n I use).
 
I am trying 3 degrees to see if it makes a difference, do you know if adding defocus to the template makes a big difference? Also, I am low-pass filtering to 40A, is that too much?

These are also important since I have to move on to smaller subunits of a ribosome, is it worth it to go lower in pixel size in the tomogram?

It would be a big help if we could use a tomogram mask on the template-matching step instead of extraction. Since we make masks and they are ~1/5 of the tomogram, it could make the 3-degree sampling much more feasible. Is this something you are thinking about adding?

Thank you for all the help,
Hamid

**Comment by McHaillet** [comment]
> I am trying 3 degrees to see if it makes a difference, do you know if adding defocus to the template makes a big difference? Also, I am low-pass filtering to 40A, is that too much?

Definitely apply a CTF that always works better. Even if its just a rough estimate. In the paper I used a single defocus for the template for all tomograms, while tilt-series always are collected with varying defocus. And for the ribosome that already is fine. The modulation of the CTF at the low frequencies is very important in that case. 

I would for the ribosomes expect quite good results (although it can of course depend on the background / tomorgam thickness), so I am not exactly what could be the issue for you. Could it be that the template is mirrored compared to the tomorgam?

Otherwise I am also happy to have a call if you want to have a look together. (you can send me a mail at m.l.chaillet@uu.nl to arrange)

**Comment by hamid13r** [comment]
Thank you very much. So I went back and just lowered the sensitivity by setting the --number-of-false-positives to 1000! and I got ~300 ribosomes back and immediately got high resolution. 

It seems like I was overestimating the number of ribosomes per tomogram. But the scoring is still a problem.

I am going to make new templates with CTF applied and see how that goes. Then I will reach out to have a talk over my data so I can show you my results more clearly.
